libraries/lib-audio-io/AudioIO.cpp

```diff
@@ -2015,104 +2023,208 @@ bool AudioIO::ProcessPlaybackSlices(
+      int bufferIndex = 0;
+      for(const auto& seq : mPlaybackSequences)
+      {
+         if(!seq)
+            continue;//no similar check in convert-to-float part
+         const auto channelGroup = seq->FindChannelGroup();
+         if(!channelGroup)
+            continue;
 
          // Are there more output device channels than channels of vt?
          // Such as when a mono sequence is processed for stereo play?
          // Then supply some non-null fake input buffers, because the
          // various ProcessBlock overrides of effects may crash without it.
          // But it would be good to find the fixes to make this unnecessary.
-         float **scratch = &mScratchPointers[mNumPlaybackChannels + 1];
-         while (iChannel < mNumPlaybackChannels)
-            memset((pointers[iChannel++] = *scratch++), 0, len * sizeof(float));
-
-         if (len && pScope) {
-            auto discardable = pScope->Process(*pGroup, &pointers[0],
-               mScratchPointers.data(),
-               // The single dummy output buffer:
-               mScratchPointers[mNumPlaybackChannels],
-               mNumPlaybackChannels, len);
-            iChannel = 0;
-            for (; iChannel < nChannels; ++iChannel) {
-               auto &ringBuffer = *mPlaybackBuffers[iBuffer + iChannel];
-               auto discarded = ringBuffer.Unput(discardable);
-               // assert(discarded == discardable);
+         auto scratch = &mScratchPointers[mNumPlaybackChannels + 1];
+
+         const auto len = mProcessingBuffers[bufferIndex].size();
+         for(unsigned i = 0; i < seq->NChannels(); ++i)
+            pointers[i] = mProcessingBuffers[bufferIndex + i].data();
+         
+         for(unsigned i = seq->NChannels(); i < mNumPlaybackChannels; ++i)
+         {
+            pointers[i] = *scratch++;
+            std::fill_n(pointers[i], mProcessingBuffers[bufferIndex].size(), .0f);
+         }
+
+         const auto discardable = pScope->Process(channelGroup, &pointers[0],
+            mScratchPointers.data(),
+            // The single dummy output buffer:
+            mScratchPointers[mNumPlaybackChannels],
+            mNumPlaybackChannels, len);
+         for(int i = 0; i < seq->NChannels(); ++i)
+         {
+            auto& buffer = mProcessingBuffers[bufferIndex + i];
+            buffer.erase(buffer.begin(), buffer.begin() + discardable);
+         }
+
+         bufferIndex += seq->NChannels();
+      }
+   }
+
+   //samples at the beginning of could have been discarded
+   //in the previous step, proceed with the number of samples
+   //equal to the shortest buffer size available
+   auto samplesAvailable = std::min_element(
+      mProcessingBuffers.begin(),
+      mProcessingBuffers.end(),
+      [](auto& first, auto& second) { return first.size() < second.size(); }
+   )->size();
+
+   //introduced latency may be too high...
+   //but we don't discard what already was written
+   if(samplesAvailable == 0)
+      return progress;
```

Comment:
I am reminded that the handling of latency is still not satisfactory.  I had ideas for fixing it that I have not found time to implement.